### PR TITLE
CORDA-1662 - Corda Serialization Evolution breaksdown with Java classes

### PR DESCRIPTION
**Pull request by fenryka**
Nullability logic was relying on annotations that Kotlin applies by
default but is left to the developer in Javaland. Change this around
so it works for both.

In Kotlin, the property must be nullable, in Java, it can't be a
primitive.

### CODE REVIEW

**Comment by mikehearn** [comment]
Is this logic really right? Why does it need to be Kotlin specific? Isn't the rule:

* If there are no annotations, it's nullable.
* If there's a NonNull annotation, it's not nullable.
* If there's a package level ParametersAreNonnullByDefault annotation, it's non-nullable unless annotated as Nullable

AFAIK the Kotlin compiler does annotate everything with nullability annotations, so we should just be able to use them.

**Comment by fenryka** [comment]
I can re-write the reflection in a more Java centric way.... which probably isn't a bad thing give nwhose going to be using it most

**Comment by fenryka** [comment]
I can't think of a nicer way of writing this than exists...

Kotlin doesn't apply specific nullability annotations to types, it embeds that info in the kotlin meta data annotation so you'd need to do a specific heck for Kotlin anyway... and you'd also need to check for primitives and java

Either a class is a Java one and it's nullable and may or may not have an annotation applied, but even if it's annotated as nullable it can be set as null and it may be a primitive type. Note, the logic here isn't that someone could feasibly add a nullable primitive, but that in a generic way the serializer can try to deserialize an older instance of the state with that new value missing. If it's nullable then we can continue as ever, if it's not or it's a primitive we have to error and be able to decide that in a generic way.

Then, on the flip side, if it is a Kotlin object we can rely on the metadata helper returning what we want

I can re-write this a number of ways but I don't see it getting any cleaner

**Comment by mikehearn** [comment]
I'm still confused, sorry.

The rule "if there's no annotation it's nullable" seems to cover the Java case.

**Comment by mikehearn** [comment]
Also Kotlin does use annotations. Look at the "Show Kotlin Bytecode" window. E.g. here's a c'tor that has a nullable parameter:

```
  public <init>(Lnet/corda/core/schemas/PersistentStateRef;)V
    @Lorg/jetbrains/annotations/Nullable;() // invisible, parameter 0
```

and one that has a non-nullable parameter:

```
  public <init>(Lnet/corda/core/contracts/StateRef;)V
    @Lorg/jetbrains/annotations/NotNull;() // invisible, parameter 0
```

**Comment by fenryka** [comment]
@mikehearn   Then reflection is working oddly as somethign like

`constructor.javaConstructor?.parameters?.forEach {
                println ("  * ${it.name} - ${it.type}")
                it.annotations?.forEach {
                    println ("    * $it")
                }
            }`

When reflecting on a Java class will print out any nullable annotations applied: such as

`static class N1 {
        private String word;
        private @Nullable Integer wibble;

        public N1(String word, @Nullable Integer wibble) {
            this.word = word;
            this.wibble = wibble;
        }
}`

Where as it won't for a Kotlin class

**Comment by fenryka** [comment]
@mikehearn the Kotlin annotations aren't retained at runtime

**Comment by shamsasari** [comment]
```

**Comment by shamsasari** [comment]
```
data class TestData(val a: String?, val b: String)

fun main(args: Array<String>) {
    val ctor = TestData::class.java.constructors[0]
    ctor.parameters.forEach {
        println("${it.name} ${it.declaredAnnotations.joinToString()}")
    }
}
```
prints just
```
a 
b 
```

Examining with javap we get
```
  public net.corda.serialization.internal.amqp.TestData(java.lang.String, java.lang.String);
    descriptor: (Ljava/lang/String;Ljava/lang/String;)V
    flags: ACC_PUBLIC
    Code:
      stack=2, locals=3, args_size=3
         0: aload_2
         1: ldc           #21                 // String b
         3: invokestatic  #27                 // Method kotlin/jvm/internal/Intrinsics.checkParameterIsNotNull:(Ljava/lang/Object;Ljava/lang/String;)V
         6: aload_0
         7: invokespecial #30                 // Method java/lang/Object."<init>":()V
        10: aload_0
        11: aload_1
        12: putfield      #11                 // Field a:Ljava/lang/String;
        15: aload_0
        16: aload_2
        17: putfield      #18                 // Field b:Ljava/lang/String;
        20: return
      LocalVariableTable:
        Start  Length  Slot  Name   Signature
            0      21     0  this   Lnet/corda/serialization/internal/amqp/TestData;
            0      21     1     a   Ljava/lang/String;
            0      21     2     b   Ljava/lang/String;
      LineNumberTable:
        line 3: 6
    MethodParameters:
      Name                           Flags
      a
      b
    RuntimeInvisibleParameterAnnotations:
      0:
        0: #7()
      1:
        0: #15()
```
```
   #7 = Utf8               Lorg/jetbrains/annotations/Nullable;
  #15 = Utf8               Lorg/jetbrains/annotations/NotNull;
```

So yes, the kotlin compiler is not paying nice with annotation visibility.

File: serialization/src/main/kotlin/net/corda/serialization/internal/amqp/EvolutionSerializer.kt [hunk 1]
```diff
@@ -117,12 +118,25 @@ abstract class EvolutionSerializer(
...
+            // case it can't be referenced
+            val isKotlin = (new.type.javaClass.declaredAnnotations.any {
+                        it.annotationClass.qualifiedName == "kotlin.Metadata"
+            })
+
             constructor.parameters.withIndex().forEach {
                 readersAsSerialized[it.value.name!!]?.apply {
                     this.resultsIndex = it.index
-                } ?: if (!it.value.type.isMarkedNullable) {
+                } ?: // If there is no value in the byte stream to map to the parameter of the constructor
```
Comment (shamsasari) [inline serialization/src/main/kotlin/net/corda/serialization/internal/amqp/EvolutionSerializer.kt]: This is difficult to make sense of. Can you convert the elvis to a normal if-statement

**Comment by fenryka** [comment]
@shamsasari ... good stuff... the alternative is what I did and read the Kotlin annotation code, you can see they've set it to CLASS and not RUNTIME

Which is vexing

**Comment by fenryka** [comment]
```Kotlin
@Documented
@Retention(RetentionPolicy.CLASS)
@Target({ElementType.METHOD, ElementType.FIELD, ElementType.PARAMETER, ElementType.LOCAL_VARIABLE})
public @interface Nullable {
  String value() default "";
}
```

File: serialization/src/test/java/net/corda/serialization/internal/amqp/JavaEvolutionTests.java [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+        SerializedBytes<Object> bytes = new SerializationOutput(factory).serialize(
+                n, TestSerializationContext.testSerializationContext);
+
+        new FileOutputStream(new File(getPath(resource).toUri())).write(bytes.getBytes());
+        */
+
+        N1 n2 = new DeserializationInput(factory).deserialize(
+                new SerializedBytes<>(Files.readAllBytes(
+                        Paths.get(this.getClass().getResource(resource).getPath()))),
```
Comment (shamsasari) [inline serialization/src/test/java/net/corda/serialization/internal/amqp/JavaEvolutionTests.java]: If you rebase onto master you'll have access to `writeTestResource()` and `readTestResource()`, which will remove all of the boilerplate.

File: serialization/src/test/java/net/corda/serialization/internal/amqp/JavaEvolutionTests.java [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+        // Uncomment to regenerate the base state of the test
+        /*
+        N2 n = new N2(w);
+
+        SerializedBytes<Object> bytes = new SerializationOutput(factory).serialize(
+                n, TestSerializationContext.testSerializationContext);
+
+        new FileOutputStream(new File(getPath(resource).toUri())).write(bytes.getBytes());
+        */
+        Assertions.assertThatThrownBy(() -> new DeserializationInput(factory).deserialize(
```
Comment (shamsasari) [inline serialization/src/test/java/net/corda/serialization/internal/amqp/JavaEvolutionTests.java]: Static import `assertThatThrownBy`

**Comment by shamsasari** [comment]
Or just look at the Kotlin code!

File: serialization/src/main/kotlin/net/corda/serialization/internal/amqp/EvolutionSerializer.kt (hunk 1 above)
Comment (fenryka) [reply serialization/src/main/kotlin/net/corda/serialization/internal/amqp/EvolutionSerializer.kt]: done

File: serialization/src/test/java/net/corda/serialization/internal/amqp/JavaEvolutionTests.java (hunk 2 above)
Comment (fenryka) [reply serialization/src/test/java/net/corda/serialization/internal/amqp/JavaEvolutionTests.java]: Hopefully, they're callable from Java in a nice way....

File: serialization/src/main/kotlin/net/corda/serialization/internal/amqp/SerializationHelper.kt [hunk 4]
```diff
@@ -540,3 +540,18 @@ fun hasCordaSerializable(type: Class<*>): Boolean {
             || type.interfaces.any(::hasCordaSerializable)
             || (type.superclass != null && hasCordaSerializable(type.superclass))
 }
+
+fun isJavaPrimitive(type: Class<*>) = type in JavaPrimitiveTypes.primativeTypes
+
+private object JavaPrimitiveTypes {
+    val primativeTypes = hashSetOf<Class<*>>(
```
Comment (shamsasari) [inline serialization/src/main/kotlin/net/corda/serialization/internal/amqp/SerializationHelper.kt]: Can't we use `Class.isPrimitive`?

File: serialization/src/test/java/net/corda/serialization/internal/amqp/JavaEvolutionTests.java [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+        public N2(String word, float wibble) {
+            this.word = word;
+            this.wibble = wibble;
+        }
+        public String getWord() { return word; }
+        public float getWibble() { return wibble; }
+    }
+
+    SerializerFactory factory = new SerializerFactory(
```
Comment (shamsasari) [inline serialization/src/test/java/net/corda/serialization/internal/amqp/JavaEvolutionTests.java]: Use testDefaultFactory instead

File: serialization/src/main/kotlin/net/corda/serialization/internal/amqp/SerializationHelper.kt (hunk 4 above)
Comment (fenryka) [reply serialization/src/main/kotlin/net/corda/serialization/internal/amqp/SerializationHelper.kt]: that's really an AMQP primitive vs a Java primtive.... the main difference is String

Comment (shamsasari) [reply serialization/src/main/kotlin/net/corda/serialization/internal/amqp/SerializationHelper.kt]: It is Java primitives: https://docs.oracle.com/javase/8/docs/api/java/lang/Class.html#isPrimitive--

File: serialization/src/test/java/net/corda/serialization/internal/amqp/JavaEvolutionTests.java [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+import java.io.*;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNull;
+
+public class JavaEvolutionTests {
+    @Rule
+    public final ExpectedException exception = ExpectedException.none();
+
+    // Class as it was when it was serialized and written to disk. Uncomment
+    // if the test referencing the object needs regeneraThrowableting.
```
Comment (shamsasari) [inline serialization/src/test/java/net/corda/serialization/internal/amqp/JavaEvolutionTests.java]: regeneraThrowableting

**Pull request merged**